cmd/zpool/zpool_main.c

```diff
@@ -2198,45 +2249,53 @@ do_import(nvlist_t *config, const char *newname, const char *mntopts,
 	uint64_t state;
 	uint64_t version;
 
-	verify(nvlist_lookup_string(config, ZPOOL_CONFIG_POOL_NAME,
-	    &name) == 0);
+	name = fnvlist_lookup_string(config, ZPOOL_CONFIG_POOL_NAME);
+	state = fnvlist_lookup_uint64(config, ZPOOL_CONFIG_POOL_STATE);
+	version = fnvlist_lookup_uint64(config, ZPOOL_CONFIG_VERSION);
 
-	verify(nvlist_lookup_uint64(config,
-	    ZPOOL_CONFIG_POOL_STATE, &state) == 0);
-	verify(nvlist_lookup_uint64(config,
-	    ZPOOL_CONFIG_VERSION, &version) == 0);
 	if (!SPA_VERSION_IS_SUPPORTED(version)) {
 		(void) fprintf(stderr, gettext("cannot import '%s': pool "
 		    "is formatted using an unsupported ZFS version\n"), name);
 		return (1);
-	} else if (state != POOL_STATE_EXPORTED &&
+	} else if (zfs_force_import_required(config) &&
 	    !(flags & ZFS_IMPORT_ANY_HOST)) {
 		uint64_t hostid = 0;
-		unsigned long system_hostid = get_system_hostid();
+		char *hostname = NULL;
+		char *import_hostname = NULL;
 
-		(void) nvlist_lookup_uint64(config, ZPOOL_CONFIG_HOSTID,
-		    &hostid);
+		hostname = fnvlist_lookup_string(config, ZPOOL_CONFIG_HOSTNAME);
 
-		if (hostid != 0 && (unsigned long)hostid != system_hostid) {
-			char *hostname;
-			uint64_t timestamp;
-			time_t t;
+		if (nvlist_exists(config, ZPOOL_CONFIG_IMPORT_HOSTNAME))
+			import_hostname = fnvlist_lookup_string(config,
+			    ZPOOL_CONFIG_IMPORT_HOSTNAME);
 
-			verify(nvlist_lookup_string(config,
-			    ZPOOL_CONFIG_HOSTNAME, &hostname) == 0);
-			verify(nvlist_lookup_uint64(config,
-			    ZPOOL_CONFIG_TIMESTAMP, &timestamp) == 0);
-			t = timestamp;
-			(void) fprintf(stderr, gettext("cannot import "
-			    "'%s': pool may be in use from other "
-			    "system, it was last accessed by %s "
-			    "(hostid: 0x%lx) on %s"), name, hostname,
-			    (unsigned long)hostid,
-			    asctime(localtime(&t)));
-			(void) fprintf(stderr, gettext("use '-f' to "
-			    "import anyway\n"));
-			return (1);
+		if (import_hostname) {
+			if (nvlist_exists(config, ZPOOL_CONFIG_IMPORT_HOSTID))
+				hostid = fnvlist_lookup_uint64(config,
+				    ZPOOL_CONFIG_IMPORT_HOSTID);
+
+			(void) fprintf(stderr, gettext("Cannot import '%s': "
+			    "pool is currently imported from %s (hostid: "
+			    "0x%lx).\nExport the pool on the other system "
+			    "before importing it with 'zpool import -f'.\n"),
+			    name, import_hostname, (unsigned long) hostid);
+		} else {
+			uint64_t timestamp = fnvlist_lookup_uint64(config,
+			    ZPOOL_CONFIG_TIMESTAMP);
+
+			if (nvlist_exists(config, ZPOOL_CONFIG_HOSTID))
+				hostid = fnvlist_lookup_uint64(config,
+				    ZPOOL_CONFIG_HOSTID);
+
+			(void) fprintf(stderr, gettext("Cannot import '%s': "
+			    "pool was previously in use from another system.\n"
+			    "Last accessed by %s (hostid: 0x%lx) at %s"
+			    "The pool can be imported, use 'zpool import -f' "
```

Comment:
Is the `The pool can be imported, use 'zpool import -f' to import the pool.` message really true? I'm running this on test hardware and attempting to import a pool that's imported on another host. If I perform a `zpool import -f`, I would still expect a failure because the activity check should fail. So I think the message is misleading.